pkg/internal/apis/meta/types.go

```diff
@@ -59,3 +59,18 @@ type SecretKeySelector struct {
 const (
 	TLSCAKey = "ca.crt"
 )
+
+// Describes how the ACME challenge self check behaves when it fails.
+type ACMESelfCheckFailurePolicy string
+
+const (
+	// The default failure policy. This policy will cause the request to be
+	// sent to the certificate provider ONLY if the ACME challenge self check
+	// succeeds. The request is never retried and the `Certificate` will remain
+	// `Pending` indefinitely.
+	ACMESelfCheckFailurePolicyWaitForever ACMESelfCheckFailurePolicy = "WaitForever"
+
+	// This policy will cause the ACME challenge self check response to be
+	// ignored and the request will always be sent to the certificate provider.
+	ACMESelfCheckFailurePolicyIgnore ACMESelfCheckFailurePolicy = "Ignore"
```

Comment:
It seems like it only does check one time, and therefore instantly sends the request to the certificate provider. This will be done too early, before the solver `Pod` and `Ingress' resources are ready...

Reply:
That's definitely going to be problematic, and without a self-check-like heuristic it'll also be difficult for us to workaround.

So looking at this from a use-cases PoV, I see a few user stories:

1) User has some form of hair-pin NAT issue, they 'know' the self check will pass if you give it 30s to propagate

2) User has a custom ingress setup that requires special steps to check

For (2), we should allow users to 'bring their own' self check implementations, i.e. 'disable' the in-built one and have a controller perform the same action.

For (1), which is what I see this PR addressing, a simple time based heuristic may be better (i.e. try once, wait 30s, accept the challenge). That said, if we were to look at this from a world where we have (2) implemented, would it make most sense for this time based check to exist as a `failurePolicy` or as an alternative self check implementation altogether?

I'm not 100% sure either way as to how we should address this/what we should do - it'd be great to get your thoughts on it and anyone else's who needs something like this.